### PR TITLE
Add components measurements retrieval

### PR DESCRIPTION
**Pull request by JerrySentry**
### Purpose/Motivation
Add a GQL resolver to retrieve measurements of components for a given repo.

### Links to relevant tickets
https://github.com/codecov/engineering-team/issues/1360

### Legal Boilerplate

Look, I get it. The entity doing business as "Sentry" was incorporated in the State of Delaware in 2015 as Functional Software, Inc. In 2022 this entity acquired Codecov and as result Sentry is going to need some rights from me in order to utilize my contributions in this PR. So here's the deal: I retain all rights, title and interest in and to my contributions, and by keeping this boilerplate intact I confirm that Sentry can use, modify, copy, and redistribute my contributions, under Sentry's choice of terms.

### CODE REVIEW

File: graphql_api/actions/flags.py [hunk 1]
```diff
@@ -48,25 +48,11 @@ def flag_measurements(
     after: datetime,
     before: datetime,
 ) -> Mapping[int, Iterable[dict]]:
```
Comment (JerrySentry) [inline graphql_api/actions/flags.py:50]: Move this to a shared generic function to be used by both flag and components measurements

File: graphql_api/actions/components.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    component_ids: Iterable[int],
+    interval: Interval,
+    after: datetime,
+    before: datetime,
+    branch: Optional[str] = None,
+) -> Mapping[int, Iterable[dict]]:
+    return measurements_by_ids(
+        repository=repository,
+        measurable_name=MeasurementName.COMPONENT_COVERAGE.value,
+        measurable_ids=component_ids,
```
Comment (ajay-sentry) [inline graphql_api/actions/components.py:20]: Did we need to do a similar cast to str here like we did with flags? 
`measurable_ids=[str(flag_id) for flag_id in flag_ids],`

File: graphql_api/types/repository/repository.py [hunk 3]
```diff
@@ -396,3 +398,53 @@ def resolve_repository_result_type(obj, *_):
...
+@sync_to_async
+def resolve_component_measurements(
+    repository: Repository,
+    info,
+    interval: Interval,
+    before: datetime,
+    after: datetime,
+    filters: Mapping = None,
+    ordering_direction: OrderingDirection = OrderingDirection.ASC,
+    branch: str = None,
```
Comment (ajay-sentry) [inline graphql_api/types/repository/repository.py:414]: nit: This should be Optional[str]

File: graphql_api/types/repository/repository.py [hunk 4]
```diff
@@ -396,3 +398,53 @@ def resolve_repository_result_type(obj, *_):
...
+@repository_bindable.field("componentMeasurements")
+@convert_kwargs_to_snake_case
+@sync_to_async
+def resolve_component_measurements(
+    repository: Repository,
+    info,
+    interval: Interval,
+    before: datetime,
+    after: datetime,
+    filters: Mapping = None,
```
Comment (ajay-sentry) [inline graphql_api/types/repository/repository.py:412]: nit: This should be Optional[Mapping]

File: services/components.py [hunk 5]
```diff
@@ -72,3 +75,41 @@ def head_totals(self) -> ReportTotals:
...
+        self.after = after
+        self.before = before
+
+    @cached_property
+    def name(self):
+        return self.component_id
+
+    @cached_property
+    def percent_covered(self):
+        if len(self.raw_measurements) > 1:
```
Comment (ajay-sentry) [inline services/components.py:101]: do we need to worry about return values for percent_covered or percent_changed if raw_measurements is 1 or 0?

Comment (JerrySentry) [reply services/components.py]: Yes, good catch, it should be ` if raw_measurements > 0`

File: graphql_api/types/repository/repository.py (hunk 4 above)
Comment (JerrySentry) [reply graphql_api/types/repository/repository.py]: Done

File: graphql_api/types/repository/repository.py (hunk 3 above)
Comment (JerrySentry) [reply graphql_api/types/repository/repository.py]: Done

File: graphql_api/actions/components.py (hunk 2 above)
Comment (JerrySentry) [reply graphql_api/actions/components.py:20]: No `measurable_ids` takes Iterable[str] which is what the type of component_ids is. Updated the `component_ids` param type to correct value.

**Review by ajay-sentry — APPROVED** [review]
lgtm!

**Pull request merged**